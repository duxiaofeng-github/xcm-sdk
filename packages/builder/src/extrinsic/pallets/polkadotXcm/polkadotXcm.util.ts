/* eslint-disable sort-keys */
import { SubmittableExtrinsicFunction } from '@polkadot/api/types';
import {
  ExtrinsicConfigBuilderPrams,
  Parents,
} from '../../ExtrinsicBuilder.interfaces';
import {
  getExtrinsicAccount,
  getExtrinsicArgumentVersion,
} from '../../ExtrinsicBuilder.utils';

export interface GetExtrinsicParams extends ExtrinsicConfigBuilderPrams {
  asset: any;
  func?: SubmittableExtrinsicFunction<'promise'>;
  parents?: Parents;
}

export function getPolkadotXcmExtrinsicArgs({
  asset,
  address,
  destination,
  func,
  parents = 1,
}: GetExtrinsicParams): any[] {
  // eslint-disable-next-line @typescript-eslint/no-use-before-define
  const version = getExtrinsicArgumentVersion(func);

  return [
    {
      [version]: {
        parents,
        interior: {
          X1: {
            Parachain: destination.parachainId,
          },
        },
      },
    },
    {
      [version]: {
        parents: 0,
        interior: {
          X1: getExtrinsicAccount(address),
        },
      },
    },
    {
      [version]: asset,
    },
    0,
    'Unlimited',
  ];
}
