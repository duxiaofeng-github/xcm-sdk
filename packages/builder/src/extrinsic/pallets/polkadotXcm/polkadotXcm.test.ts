import {
  buildParachainParamsMock,
  buildParamsMock,
} from '../../../../fixtures';
import { polkadotXcm } from './polkadotXcm';

describe('polkadotXcm', () => {
  describe('limitedReserveTransferAssets', () => {
    describe('here', () => {
      const extrinsic = polkadotXcm()
        .limitedReserveTransferAssets()
        .here()
        .build(buildParamsMock);

      it('should be correct config', () => {
        expect(extrinsic).toMatchSnapshot();
      });

      it('should get correct arguments', () => {
        expect(extrinsic.getArgs()).toMatchSnapshot();
      });
    });

    describe('x1', () => {
      const extrinsic = polkadotXcm()
        .limitedReserveTransferAssets()
        .X1()
        .build(buildParamsMock);

      it('should be correct config', () => {
        expect(extrinsic).toMatchSnapshot();
      });

      it('should get correct arguments', () => {
        expect(extrinsic.getArgs()).toMatchSnapshot();
      });
    });

    describe('x2', () => {
      const extrinsic = polkadotXcm()
        .limitedReserveTransferAssets()
        .X2()
        .build(buildParamsMock);

      it('should be correct config', () => {
        expect(extrinsic).toMatchSnapshot();
      });

      it('should get correct arguments', () => {
        expect(extrinsic.getArgs()).toMatchSnapshot();
      });
    });

    describe('parachain', () => {
      const extrinsic = polkadotXcm()
        .limitedReserveTransferAssets()
        .X2()
        .build(buildParachainParamsMock);

      it('should be correct config', () => {
        expect(extrinsic).toMatchSnapshot();
      });

      it('should get correct arguments', () => {
        expect(extrinsic.getArgs()).toMatchSnapshot();
      });
    });
  });

  describe('limitedReserveWithdrawAssets', () => {
    describe('x2', () => {
      const extrinsic = polkadotXcm()
        .limitedReserveWithdrawAssets()
        .X2()
        .build(buildParamsMock);

      it('should be correct config', () => {
        expect(extrinsic).toMatchSnapshot();
      });

      it('should get correct arguments', () => {
        expect(extrinsic.getArgs()).toMatchSnapshot();
      });
    });

    describe('parachain', () => {
      const extrinsic = polkadotXcm()
        .limitedReserveWithdrawAssets()
        .X2()
        .build(buildParachainParamsMock);

      it('should be correct config', () => {
        expect(extrinsic).toMatchSnapshot();
      });

      it('should get correct arguments', () => {
        expect(extrinsic.getArgs()).toMatchSnapshot();
      });
    });
  });
});
