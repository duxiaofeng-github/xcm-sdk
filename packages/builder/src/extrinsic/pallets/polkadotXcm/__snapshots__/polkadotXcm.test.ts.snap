// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`polkadotXcm limitedReserveTransferAssets here should be correct config 1`] = `
ExtrinsicConfig {
  "func": "limitedReserveTransferAssets",
  "getArgs": [Function],
  "module": "polkadotXcm",
  "type": "Substrate",
}
`;

exports[`polkadotXcm limitedReserveTransferAssets here should get correct arguments 1`] = `
[
  {
    "V1": {
      "interior": {
        "X1": {
          "Parachain": 1000,
        },
      },
      "parents": 1,
    },
  },
  {
    "V1": {
      "interior": {
        "X1": {
          "AccountKey20": {
            "key": "0xeF46c7649270C912704fB09B75097f6E32208b85",
            "network": "Any",
          },
        },
      },
      "parents": 0,
    },
  },
  {
    "V1": [
      {
        "fun": {
          "Fungible": 99000000000n,
        },
        "id": {
          "Concrete": {
            "interior": "Here",
            "parents": 0,
          },
        },
      },
    ],
  },
  0,
  "Unlimited",
]
`;

exports[`polkadotXcm limitedReserveTransferAssets parachain should be correct config 1`] = `
ExtrinsicConfig {
  "func": "limitedReserveTransferAssets",
  "getArgs": [Function],
  "module": "polkadotXcm",
  "type": "Substrate",
}
`;

exports[`polkadotXcm limitedReserveTransferAssets parachain should get correct arguments 1`] = `
[
  {
    "V1": {
      "interior": {
        "X1": {
          "Parachain": 2032,
        },
      },
      "parents": 1,
    },
  },
  {
    "V1": {
      "interior": {
        "X1": {
          "AccountId32": {
            "id": "0x18e6c7ad45fbab18778710642072fdd057470b76d4d568b44dea6accd5b0c423",
            "network": null,
          },
        },
      },
      "parents": 0,
    },
  },
  {
    "V1": [
      {
        "fun": {
          "Fungible": 99000000000n,
        },
        "id": {
          "Concrete": {
            "interior": {
              "X2": [
                {
                  "PalletInstance": 10,
                },
                {
                  "GeneralIndex": "USDT",
                },
              ],
            },
            "parents": 0,
          },
        },
      },
    ],
  },
  0,
  "Unlimited",
]
`;

exports[`polkadotXcm limitedReserveTransferAssets x1 should be correct config 1`] = `
ExtrinsicConfig {
  "func": "limitedReserveTransferAssets",
  "getArgs": [Function],
  "module": "polkadotXcm",
  "type": "Substrate",
}
`;

exports[`polkadotXcm limitedReserveTransferAssets x1 should get correct arguments 1`] = `
[
  {
    "V1": {
      "interior": {
        "X1": {
          "Parachain": 1000,
        },
      },
      "parents": 1,
    },
  },
  {
    "V1": {
      "interior": {
        "X1": {
          "AccountKey20": {
            "key": "0xeF46c7649270C912704fB09B75097f6E32208b85",
            "network": "Any",
          },
        },
      },
      "parents": 0,
    },
  },
  {
    "V1": [
      {
        "fun": {
          "Fungible": 99000000000n,
        },
        "id": {
          "Concrete": {
            "interior": {
              "X1": {
                "PalletInstance": 10,
              },
            },
            "parents": 0,
          },
        },
      },
    ],
  },
  0,
  "Unlimited",
]
`;

exports[`polkadotXcm limitedReserveTransferAssets x2 should be correct config 1`] = `
ExtrinsicConfig {
  "func": "limitedReserveTransferAssets",
  "getArgs": [Function],
  "module": "polkadotXcm",
  "type": "Substrate",
}
`;

exports[`polkadotXcm limitedReserveTransferAssets x2 should get correct arguments 1`] = `
[
  {
    "V1": {
      "interior": {
        "X1": {
          "Parachain": 1000,
        },
      },
      "parents": 1,
    },
  },
  {
    "V1": {
      "interior": {
        "X1": {
          "AccountKey20": {
            "key": "0xeF46c7649270C912704fB09B75097f6E32208b85",
            "network": "Any",
          },
        },
      },
      "parents": 0,
    },
  },
  {
    "V1": [
      {
        "fun": {
          "Fungible": 99000000000n,
        },
        "id": {
          "Concrete": {
            "interior": {
              "X2": [
                {
                  "PalletInstance": 10,
                },
                {
                  "GeneralIndex": "USDT",
                },
              ],
            },
            "parents": 0,
          },
        },
      },
    ],
  },
  0,
  "Unlimited",
]
`;

exports[`polkadotXcm limitedReserveWithdrawAssets parachain should be correct config 1`] = `
ExtrinsicConfig {
  "func": "limitedReserveWithdrawAssets",
  "getArgs": [Function],
  "module": "polkadotXcm",
  "type": "Substrate",
}
`;

exports[`polkadotXcm limitedReserveWithdrawAssets parachain should get correct arguments 1`] = `
[
  {
    "V1": {
      "interior": {
        "X1": {
          "Parachain": 2032,
        },
      },
      "parents": 1,
    },
  },
  {
    "V1": {
      "interior": {
        "X1": {
          "AccountId32": {
            "id": "0x18e6c7ad45fbab18778710642072fdd057470b76d4d568b44dea6accd5b0c423",
            "network": null,
          },
        },
      },
      "parents": 0,
    },
  },
  {
    "V1": [
      {
        "fun": {
          "Fungible": 99000000000n,
        },
        "id": {
          "Concrete": {
            "interior": {
              "X2": [
                {
                  "Parachain": 2032,
                },
                {
                  "PalletInstance": 10,
                },
              ],
            },
            "parents": 1,
          },
        },
      },
    ],
  },
  0,
  "Unlimited",
]
`;

exports[`polkadotXcm limitedReserveWithdrawAssets x2 should be correct config 1`] = `
ExtrinsicConfig {
  "func": "limitedReserveWithdrawAssets",
  "getArgs": [Function],
  "module": "polkadotXcm",
  "type": "Substrate",
}
`;

exports[`polkadotXcm limitedReserveWithdrawAssets x2 should get correct arguments 1`] = `
[
  {
    "V1": {
      "interior": {
        "X1": {
          "Parachain": 1000,
        },
      },
      "parents": 1,
    },
  },
  {
    "V1": {
      "interior": {
        "X1": {
          "AccountKey20": {
            "key": "0xeF46c7649270C912704fB09B75097f6E32208b85",
            "network": "Any",
          },
        },
      },
      "parents": 0,
    },
  },
  {
    "V1": [
      {
        "fun": {
          "Fungible": 99000000000n,
        },
        "id": {
          "Concrete": {
            "interior": {
              "X2": [
                {
                  "Parachain": 1000,
                },
                {
                  "PalletInstance": 10,
                },
              ],
            },
            "parents": 1,
          },
        },
      },
    ],
  },
  0,
  "Unlimited",
]
`;
